source/common/tls/server_context_impl.cc

```diff
@@ -652,45 +653,58 @@ ServerContextImpl::findTlsContext(absl::string_view sni, bool client_ecdsa_capab
 
 enum ssl_select_cert_result_t
 ServerContextImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello) {
-  absl::string_view sni = absl::NullSafeStringView(
-      SSL_get_servername(ssl_client_hello->ssl, TLSEXT_NAMETYPE_host_name));
-  const bool client_ecdsa_capable = isClientEcdsaCapable(ssl_client_hello);
-  const bool client_ocsp_capable = isClientOcspCapable(ssl_client_hello);
-
-  auto [selected_ctx, ocsp_staple_action] =
-      findTlsContext(sni, client_ecdsa_capable, client_ocsp_capable, nullptr);
-
-  // Apply the selected context. This must be done before OCSP stapling below
-  // since applying the context can remove the previously-set OCSP response.
-  // This will only return NULL if memory allocation fails.
-  RELEASE_ASSERT(SSL_set_SSL_CTX(ssl_client_hello->ssl, selected_ctx.ssl_ctx_.get()) != nullptr,
-                 "");
-
-  if (client_ocsp_capable) {
-    stats_.ocsp_staple_requests_.inc();
+  if (tls_certificate_selector_ == nullptr) {
+    // Lazy init the provider here since we can not use shared_from_this in construct method.
+    tls_certificate_selector_ = tls_certificate_selector_factory_cb_(shared_from_this());
   }
+  ASSERT(tls_certificate_selector_ != nullptr);
 
-  switch (ocsp_staple_action) {
-  case OcspStapleAction::Staple: {
-    // We avoid setting the OCSP response if the client didn't request it, but doing so is safe.
-    RELEASE_ASSERT(selected_ctx.ocsp_response_,
-                   "OCSP response must be present under OcspStapleAction::Staple");
-    auto& resp_bytes = selected_ctx.ocsp_response_->rawBytes();
-    int rc = SSL_set_ocsp_response(ssl_client_hello->ssl, resp_bytes.data(), resp_bytes.size());
-    RELEASE_ASSERT(rc != 0, "");
-    stats_.ocsp_staple_responses_.inc();
-  } break;
-  case OcspStapleAction::NoStaple:
-    stats_.ocsp_staple_omitted_.inc();
+  auto* extended_socket_info = reinterpret_cast<Envoy::Ssl::SslExtendedSocketInfo*>(
+      SSL_get_ex_data(ssl_client_hello->ssl, ContextImpl::sslExtendedSocketInfoIndex()));
+
+  auto selection_result = extended_socket_info->certSelectionResult();
+  switch (selection_result) {
+  case Ssl::CertSelectionStatus::NotStarted:
+    // continue
     break;
-  case OcspStapleAction::Fail:
-    stats_.ocsp_staple_failed_.inc();
+
+  case Ssl::CertSelectionStatus::Pending:
```

Comment:
Why would this function get called if it's already pending?

Reply:
`selectTlsContext` will be invoked again, after it returning `ssl_select_cert_retry`, but I'm not clear about why it happens.
After remove these three lines, `tls_certificate_selector_test` will fail.